llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -1908,6 +1909,8 @@ SDValue DAGCombiner::visit(SDNode *N) {
   case ISD::ZERO_EXTEND_VECTOR_INREG:
   case ISD::ANY_EXTEND_VECTOR_INREG: return visitEXTEND_VECTOR_INREG(N);
   case ISD::TRUNCATE:           return visitTRUNCATE(N);
+  case ISD::TRUNCATE_SSAT_U:
+  case ISD::TRUNCATE_USAT_U:    return visitTRUNCATE_USAT(N);
```

Comment:
Should this be called visitTRUNCATE_SAT_U? Or should we just have a visitTRUNCATE_SAT call and handle the unsigned cases inside it?

Reply:
Did we not mean to change this to just `case ISD::TRUNCATE_USAT_U: return visitTRUNCATE_USAT(N);`, without the TRUNCATE_SSAT_U case? (Sorry if I missed that). Otherwise it will change `TRUNCATE_SSAT_U(FP_TO_UINT(x))` to `FP_TO_UINT_SAT(x)`, which will not clamp to the same bounds.

Reply:
visitTRUNCATE_USAT has a small task, so I don't see the need to separate SSAT and USAT.
Is it LLVM's way to reduce unnecessary separation and separate them if they are reasonably large?

Reply:
No, we can generalize it later if the need arises - but we need to confirm if we should be handling the TRUNCATE_SSAT_U  case or not (do we have test coverage?)

Reply:
If that's the case, then we don't have a test for truncate_ssat_u. As @davemgreen commented, it is right to change to call visitTRUNCATE_USAT() in case TRUNCATE_USAT_U.

Reply:
should I change it so that only ` case ISD::TRUNCATE_USAT_U: return visitTRUNCATE_USAT(N);` remains?

Reply:
I think so. Perhaps call it visitTRUNCATE_USAT_U too?